src/plonk/eval.rs

```diff
@@ -70,12 +67,34 @@ pub trait Eval<F: PrimeField> {
             },
         )
     }
+}
 
-    /// evaluate virtual multi-polynomial (i.e. custom gates, cross-term expressions etc) on specific row
-    /// to evaluate each monomial term of the form $c*x1[row]^{k1}*x2[row]^{k2}*\cdots$, we first lookup
-    /// the value of $x[row]$ from the EvaluationDomain, then calculate the value of monomial.
-    /// to speedup, we will save the x and x^k in a HashMap
-    fn eval(&self, poly: &MultiPolynomial<F>, row: usize) -> Result<F, Error> {
+/// The `Eval` trait is used to evaluate multi-variable polynomials in a Evaluation Domain defined
+/// over a prime field `F`
+///
+/// This trait encapsulates the necessary functionality to evaluate polynomials
+///
+/// # Generic Parameters
+///
+/// * `F`: The prime field over which the evaluations are carried out.
+/// * `POLY`: The type of polynomial or expression to be evaluated.
+///
+pub trait Eval<F: PrimeField, POLY>: GetDataForEval<F> {
+    /// General method allowing to perform calculations of input polynomial type on the basis of
+    /// data requested in `GetEvalData`. Depending on its representation the implementation may
+    /// differ
+    fn eval(&self, poly: &POLY, row_index: usize) -> Result<F, Error>;
+}
+
+impl<F: PrimeField, E: GetDataForEval<F>> Eval<F, MultiPolynomial<F>> for E {
```

Comment:
Since we are going to remove the MultiPolynomial at the end, after that, we don't need two generics `<F, E>`.  We can change it back later.

Reply:
The main problem with this API, then, is we have to count for one instance of `GraphEvaluator` all strings, so don't pay much attention to this particular `Eval` design, as it doesn't fit anyway. In theory, you could tweak it by making some of the
``rust
struct Evaluator {
    graph_evaluator: GraphEvaluator,
    data: impl GetDataForEval
}
```
But I will propose this design after I debug the final code.